### PR TITLE
Track related refactoring

### PR DESCRIPTION
**Pull request by AEtherC0r3**
* Add roles as nested routes to track (for the track organizer role)
* Allow transition from to_accept to to_reject and backwards
* Split Track#valid_dates validation to many independent ones
* Show all the confirmed tracks in the conference's splashpage
* Add comment in admin/Tracks#toggle_cfp_inclusion
* Rewrite admin/TracksController#accept spec
* Add feature spec for track requests
* Change 'In' to 'Room' in Tracks#index
* Rewrite Track#overlapping
* Refactor code in ProposalsController
* Fix typos

### CODE REVIEW

File: spec/controllers/admin/tracks_controller_spec.rb [hunk 1]
```diff
@@ -383,12 +376,16 @@
...
+      @room = create(:room, venue: conference.venue)
+    end
+
     context 'has start_date, end_date and room' do
       before :each do
-        self_organized_track.start_date = Date.today
-        self_organized_track.end_date = Date.today
-        conference.venue = create(:venue)
-        self_organized_track.room = create(:room, venue: conference.venue)
+        self_organized_track.start_date = Date.current
```
Comment (differentreality) [inline spec/controllers/admin/tracks_controller_spec.rb]: For this case you can simply define a variable, using let, named full_self_organized_track or something similar, that indicates that this is a self_organized track with all of its attributes.

File: spec/controllers/admin/tracks_controller_spec.rb [hunk 2]
```diff
@@ -409,31 +406,80 @@
     end
 
     context 'has start_date and end_date' do
-      it_behaves_like 'fails to accept', true, true, false
+      before :each do
+        self_organized_track.start_date = Date.current
+        self_organized_track.end_date = Date.current
```
Comment (differentreality) [inline spec/controllers/admin/tracks_controller_spec.rb]: Using the predefined track with all attributes, you can be simply removing the ones you do **not** want, which is going to reduce the lines of our tests (that's good because we have less to read and understand).

File: spec/controllers/admin/tracks_controller_spec.rb [hunk 3]
```diff
@@ -409,31 +406,80 @@
...
+      end
+
+      it_behaves_like 'fails to accept'
     end
 
-    context 'has non of start_date, end_date, room' do
-      it_behaves_like 'fails to accept', false, false, false
+    context 'has none of start_date, end_date, room' do
+      before :each do
+        self_organized_track.start_date = nil
```
Comment (differentreality) [inline spec/controllers/admin/tracks_controller_spec.rb]: We can have another variable for a track that has none of those extra attributes (dates, room) set. The same variable can be used in other cases, where you only want to set one of the attribute to non nil.

File: spec/features/tracks_spec.rb [hunk 4]
```diff
@@ -60,12 +61,83 @@
       within('table#tracks') do
         expect(page.has_content?('Distribution')).to be true
         expect(page.has_content?('Events about our Linux')).to be true
-        expect(page.assert_selector('tr', count: 2)).to be true
+        expect(Track.count).to eq 1
+      end
+    end
+  end
+
+  shared_examples 'tracks' do
```
Comment (differentreality) [inline spec/features/tracks_spec.rb]: These shared examples are all about managing / working with tracks, right?
Perhaps the name could be a little bit more specific?

File: app/models/track.rb [hunk 5]
```diff
@@ -29,8 +29,12 @@ class Track < ActiveRecord::Base
   validates :room, presence: true, if: :self_organized_and_accepted_or_confirmed?
   validates :relevance, presence: true, if: :self_organized?
   validates :description, presence: true, if: :self_organized?
-  validate :valid_dates
-  validate :valid_room, if: :self_organized_and_accepted_or_confirmed?
+  validate :start_date_after_conference_start_date
```
Comment (differentreality) [inline app/models/track.rb]: I'd combine the test related to the conference date into 1, dates_within_conference_dates or something similar, and I would test if start_date is within the range of conf.start_date..conf.end_date

As long as  you have the validation start_date_before_end_date, that should be sufficient. What do you think?

File: spec/models/track_spec.rb [hunk 6]
```diff
@@ -84,25 +84,85 @@
...
+      end
+    end
+
+    describe '#start_date_before_end_date' do
+      before :each do
+        @conference = create(:conference, start_date: 1.day.ago, end_date: 2.days.from_now)
+      end
 
+      context 'is valid' do
+        it 'when the track\'s start date is before its end date' do
```
Comment (differentreality) [inline spec/models/track_spec.rb]: This can probably be combined with https://github.com/openSUSE/osem/pull/1639/files#diff-c2212054e3536a54e738fecea403adffR117

so that we are testing track dates within conference dates (I would even have it start on conf.start_date and end on conf.end_date which is our edge case).

And that would be enough for our tests, since we will test the start_date before end_date in the other validation.

**Comment by differentreality** [comment]
I'd also transform tracks#index to show table headers for 'From', 'To', 'In'

**Comment by AEtherC0r3** [comment]
@differentreality I'm having trouble refactoring the refactored admin/TracksController spec. The variable I define (@this_track) isn't available in the `it_behaves_like` directive, to be passed as a variable. Also, I can't pass a track defined in a let to it. I'm including the commit for reference below. In my opinion, we should ditch the second track and, this way, get rid of the need to have a variable. This is a hybrid solution between my initially refactored code and what you proposed in the comments. What do you think?

File: app/views/tracks/index.html.haml [hunk 7]
```diff
@@ -29,6 +29,13 @@
     .row
       .col-md-12
         %table.table.table-striped#tracks
+          %th
+          %th
+          %th
+          %th From
+          %th To
+          %th In
```
Comment (differentreality) [inline app/views/tracks/index.html.haml]: Perhaps that's more meaningful if it says 'Room' or 'In room', what do you think?

Comment (AEtherC0r3) [inline app/views/tracks/index.html.haml]: I think that 'Room' is enough

**Comment by differentreality** [comment]
Nice work @AEtherC0r3 :+1:
